tests/test_bin_trie.py

```diff
@@ -0,0 +1,40 @@
+import pytest
+
+import random
+
+from trie.trie import (
+    BinaryTrie,
+)
+
+from trie.constants import (
+    BLANK_HASH,
+)
+from trie.utils.sha3 import (
+    keccak,
+)
+
+
+kvpairs = [(keccak(str(i).encode('utf-8')), str(i).encode('utf-8') * 5) for i in range(2000)]
+
+
+def test_bin_trie_different_order_insert():
+    result = BLANK_HASH
+    # Repeat 3 times
+    for _ in range(3):
```

Comment:
Can we change the approach to this test.  What I'm worried about is this finding an edge case that fails but not actually knowing what it was because it involves some unknown random ordering.

I suggest we do the following.

- convert to use `hypothesis` for randomness generation.

```
from hypothesis import (
    given,
    strategies as st,
)

@given(kv_pairs=st.lists(st.tuples(st.binary(), st.binary())))
def test_bin_trie_different_order_insert(kv_pairs):
    ...
```

Then, to check that ordering is unimportant we still need randomness.  `hypothesis` can help with this too via `hypothesis.strategies.random_module()` which returns the equivalent of `import random` but with the seed set by hypothesis so that the test can be repeated deterministically. 

```
from hypothesis import (
    given,
    strategies as st,
)

@given(kv_pairs=st.lists(st.tuples(st.binary(), st.binary())), random=st.random_module())
def test_bin_trie_different_order_insert(kv_pairs, random):
    ....
    random.shuffle(kv_pairs)
    ...
```

Reply:
some limits with random kv pairs are:
1. in binary trie, one key can not be the prefix of the other key, e.g, b'\x00\x01' and b'\x00\x01\x02'
2. can not have updates with different values but the same key, otherwise it might result in different root hash after shuffling

so I make some twists:
```
@to_list
def to_kv_pairs(kv_zip):
    for k, v in kv_zip:
        yield (k, v)

@given(k=st.lists(st.binary(min_size=32, max_size=32), min_size=10, max_size=10, unique=True),
    v=st.lists(st.binary(min_size=1), min_size=10, max_size=10),
    set_random=st.random_module())
def test_bin_trie_different_order_insert(k, v, set_random):
    kv_pairs = to_kv_pairs(zip(k, v))
```
Which basically make all keys to be 32 bytes long and values to be 1 bytes minimum.
Numbers including list size are all adjustable

Reply:
Looks like a solid approach. 👍